kube-runtime/src/utils/backoff_reset_timer.rs

```diff
@@ -0,0 +1,86 @@
+use std::time::{Duration, Instant};
+
+use backoff::{backoff::Backoff, Clock, SystemClock};
+
+/// A [`Backoff`] wrapper that resets after a fixed duration has elapsed.
+struct ResetTimerBackoff<B, C = SystemClock> {
+    backoff: B,
+    clock: C,
+    last_backoff: Option<Instant>,
+    reset_duration: Duration,
+}
+
+impl<B: Backoff> ResetTimerBackoff<B> {
+    fn new(backoff: B, reset_duration: Duration) -> Self {
+        Self::new_with_custom_clock(backoff, reset_duration, SystemClock {})
+    }
+}
+
+impl<B: Backoff, C: Clock> ResetTimerBackoff<B, C> {
+    fn new_with_custom_clock(backoff: B, reset_duration: Duration, clock: C) -> Self {
+        Self {
+            backoff,
+            clock,
+            last_backoff: None,
+            reset_duration,
+        }
+    }
+}
+
+impl<B: Backoff, C: Clock> Backoff for ResetTimerBackoff<B, C> {
+    fn next_backoff(&mut self) -> Option<Duration> {
+        if let Some(last_backoff) = self.last_backoff {
+            if self.clock.now() > last_backoff + self.reset_duration {
+                tracing::debug!(
+                    ?last_backoff,
+                    reset_duration = ?self.reset_duration,
+                    "Resetting backoff, since reset duration has expired"
+                );
+                self.backoff.reset();
+            }
+        }
+        self.last_backoff = Some(self.clock.now());
+        self.backoff.next_backoff()
+    }
+
+    fn reset(&mut self) {
+        // Do not even bother trying to reset here, since `next_backoff` will take care of this when the timer expires.
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use backoff::{backoff::Backoff, Clock};
+    use tokio::time::advance;
+
+    use super::ResetTimerBackoff;
+    use crate::utils::stream_backoff::tests::LinearBackoff;
```

Comment:
Cross-depending on a separate module's test crate like this feels pretty awkward...